policy-controller/k8s/status/src/index.rs

```diff
@@ -0,0 +1,246 @@
+use crate::{
+    http_route::{ParentReference, RouteBinding},
+    resource_id::ResourceId,
+};
+use ahash::{AHashMap as HashMap, AHashSet as HashSet};
+use chrono::offset::Utc;
+use linkerd_policy_controller_k8s_api::{self as k8s, gateway, ResourceExt};
+use parking_lot::RwLock;
+use std::{collections::hash_map::Entry, sync::Arc};
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+
+const POLICY_API_GROUP: &str = "policy.linkerd.io";
+const POLICY_API_VERSION: &str = "policy.linkerd.io/v1alpha1";
+const STATUS_CONTROLLER_NAME: &str = "policy.linkerd.io/status-controller";
+
+pub type SharedIndex = Arc<RwLock<Index>>;
+
+pub struct Controller {
+    client: k8s::Client,
+    index: SharedIndex,
+    updates: UnboundedReceiver<Update>,
+}
+
+pub struct Index {
+    updates: UnboundedSender<Update>,
+
+    http_routes: HashMap<ResourceId, RouteBinding>,
+    servers: HashSet<ResourceId>,
+}
+
+pub enum Update {
+    HttpRoute(ResourceId),
+    Server,
+}
+
+impl Controller {
+    pub fn new(
+        client: k8s::Client,
+        index: SharedIndex,
+        updates: UnboundedReceiver<Update>,
+    ) -> Self {
```

Comment:
Very nitpicky: not sure what the idiomatic way to use Self is in Rust, but I prefer not to use it as a return type since I always have to go back to the impl block (which I guess is not a huge problem with `new()` functions. I think @hawkw would have a better opinion here though so I'd just go with what she says. It'd be cool for my own benefit to see what we prefer to do across our Rust repos tho.

Reply:
Yea we seem to use `Self` in most places that we can in other policy controller code as well as the proxy, but I can change it if we want to.

Reply:
it's generally considered idiomatic to use `Self` as a return type for constructors, rather than the explicit type name.

one advantage of doing this is that it means that changes to the type's name or adding generic parameters don't need to be made in multiple places if `Self` is used, while the return type would need to be changed if it was written out explicitly.